src/fstar/FStar.CheckedFiles.fs

```diff
@@ -0,0 +1,335 @@
+                  BU.smap_add mcache checked_fn (Invalid msg, Inl msg)
+                end
+                else BU.smap_add mcache checked_fn (Unknown (deps_dig, dig, tc_result), Inr parsing_data)
+
+(*
+ * Second step for loading checked files, validates the tc data
+ *)
+let load_checked_file_with_tc_result (deps:Dep.deps) (fn:string) (checked_fn:string) :unit =
+  load_checked_file fn checked_fn;  //first step, in case some client calls it directly
+  match BU.smap_try_find mcache checked_fn with
+  | None -> failwith "Impossible, load_checked_file must add an entry to mcache"
+  | Some (Invalid _, _)
+  | Some (Valid _, _) -> ()  //already marked Valid or Invalid
+  | Some (Unknown (deps_dig, dig, tc_result), parsing_data) ->
+    match hash_dependences deps fn with
+    | Inl msg ->
+      BU.smap_add mcache checked_fn (Invalid msg, parsing_data)
+    | Inr deps_dig' ->
+      if deps_dig = deps_dig'
+      then begin
+        //mark the tc data of the file as valid
+        BU.smap_add mcache checked_fn (Valid (dig, tc_result), parsing_data);
+        //if there exists an interface for it, mark that too as valid
+        //validity of implementaton tc data implies validity of iface tc data
+        let validate_iface_cache () =
+          let iface = fn |> Dep.lowercase_module_name |> Dep.interface_of deps in
+          match iface with
+          | None -> ()
+          | Some iface ->
+            try
+              let iface_checked_fn = iface |> Dep.cache_file_name in
+              match BU.smap_try_find mcache iface_checked_fn with
+              | Some (Unknown (_, dig, tc_result), parsing_data) ->
+                BU.smap_add mcache iface_checked_fn (Valid (dig, tc_result), parsing_data)
+              | _ -> ()
+            with
+              | _ -> ()
+        in
+        validate_iface_cache ()
+      end
+      else begin
+        if Options.debug_any()
+        then begin
+          BU.print4 "Expected (%s) hashes:\n%s\n\nGot (%s) hashes:\n\t%s\n"
+            (BU.string_of_int (List.length deps_dig'))
+            (FStar.Parser.Dep.print_digest deps_dig')
+            (BU.string_of_int (List.length deps_dig))
+            (FStar.Parser.Dep.print_digest deps_dig);
+          if List.length deps_dig = List.length deps_dig'
+          then List.iter2 (fun (x,y) (x', y') ->
+                 if x<>x' || y<>y'
+                 then BU.print2 "Differ at: Expected %s\n Got %s\n"
+                                (FStar.Parser.Dep.print_digest [(x,y)])
+                                (FStar.Parser.Dep.print_digest [(x',y')])) deps_dig deps_dig'
+        end;
+        let msg =
+          BU.format1
+            "checked file %s is stale (dependence hash mismatch, use --debug yes for more details)"
+            checked_fn
+        in
+        BU.smap_add mcache checked_fn (Invalid msg, Inl msg)
+      end
+
+
+let load_parsing_data_from_cache file_name =
+  let cache_file =  //AR: why are we catching this exception?
```

Comment:
Yeah, this is odd. That call to cache_file_name can raise a fatal error "FStar.Errors.Error_AlreadyCachedAssertionFailure". We shouldn't suppress it.